### ISSUE TITLE
QueryParamsHandling 'merge' does not remove query parameters

### ISSUE THREAD

**Issue by mkennedy3000** [angular/angular#18463]
## I'm submitting a...

<pre><code>
[x] Bug report
</code></pre>

## Current behavior

When navigating to a route with query params and the `'merge'` queryParamHandling strategy, if a specified query param is `null` the specified query param is kept if it previously existed.

## Expected behavior

When navigating to a route with query params and the `'merge'` queryParamHandling strategy, if a specified query param is `null` the specified query param is removed if it previously existed.

## Minimal reproduction of the problem with instructions

* an application with one route `/`
* `router.navigate(['/'], { queryParams: { hello: 'world' }, queryParamHandling: 'merge' })` => `/?hello=world`
* `router.navigate(['/'], { queryParams: { hello: null }, queryParamHandling: 'merge' }` => `/?hello=world` (but expected `/`)

## What is the motivation / use case for changing the behavior?

I have a route with multiple query params, including `filters`. The `filters` are updated independently of the other query params but I don't want to wipe them out when `filters` are updated (that is why I use `'merge'`). However, I currently cannot remove a query param with `'merge'` currently and I expect to be able to because the provided value of `null` should override an existing value and remove the query param.

The current workaround I have is to use `Object.assign` and the snapshot of the previous query params with the default QueryParamHandling to get the behavior I desire.

## Environment

<pre><code>
Angular version: 4.2.3
</code></pre>

**Comment by mkennedy3000** [comment]
Related to #17202, but I consider the behavior a bug.

**Comment by skreborn** [comment]
This seems to happen because empty parameters are removed *before* the merge happens. See the [`navigate`](https://github.com/angular/angular/blob/3a227a1f6fb9d734aeba83cdf80246ae775a4f53/packages/router/src/router.ts#L461) method of the router.

```
navigate(commands: any[], extras: NavigationExtras = {skipLocationChange: false}):
      Promise<boolean> {
    validateCommands(commands);
    if (typeof extras.queryParams === 'object' && extras.queryParams !== null) {
      extras.queryParams = this.removeEmptyProps(extras.queryParams);
    }
    return this.navigateByUrl(this.createUrlTree(commands, extras), extras);
}
```

This might simply be fixed by rearranging this function a little:


```
navigate(commands: any[], extras: NavigationExtras = {skipLocationChange: false}):
      Promise<boolean> {
    validateCommands(commands);
    const urlTree = this.createUrlTree(commands, extras);
    if (typeof urlTree.queryParams === 'object' && urlTree.queryParams !== null) {
      urlTree.queryParams = this.removeEmptyProps(urlTree.queryParams);
    }
    return this.navigateByUrl(urlTree, extras);
}
```

This code has never been tested and might just be stupidly wrong, but I hope it gets the idea through.

**Comment by bossqone** [comment]
I wanted to send PR, but I don't know how to properly write test for `Router` (especially for `createUrlTree`). Solution provided by @skreborn should work, but I think we could remove empty props directly in `Router.createUrlTree`.

Here's my proposal of `Router.createUrlTree`:
```ts
createUrlTree(commands: any[], navigationExtras: NavigationExtras = {}): UrlTree {
  const {relativeTo,          queryParams,         fragment,
         preserveQueryParams, queryParamsHandling, preserveFragment} = navigationExtras;
  if (isDevMode() && preserveQueryParams && <any>console && <any>console.warn) {
    console.warn('preserveQueryParams is deprecated, use queryParamsHandling instead.');
  }
  const a = relativeTo || this.routerState.root;
  const f = preserveFragment ? this.currentUrlTree.fragment : fragment;
  let q: Params|null = null;
  if (queryParamsHandling) {
    switch (queryParamsHandling) {
      case 'merge':
        q = {...this.currentUrlTree.queryParams, ...queryParams};
        break;
      case 'preserve':
        q = this.currentUrlTree.queryParams;
        break;
      default:
        q = queryParams || null;
    }
  } else {
    q = preserveQueryParams ? this.currentUrlTree.queryParams : queryParams || null;
  }

  // HERE IS INTRODUCED CHANGE
  if (q !== null) {
    q = this.removeEmptyProps(q);
  }
  return createUrlTree(a, this.currentUrlTree, commands, q !, f !);
}
```
And then `Router.navigate` should be changed to:
```ts
navigate(commands: any[], extras: NavigationExtras = {skipLocationChange: false}):
    Promise<boolean> {
  validateCommands(commands);
  return this.navigateByUrl(this.createUrlTree(commands, extras), extras);
}
```

**Comment by bossqone** [comment]
any updates on this?

**Comment by tproenca** [comment]
Any updates of this?

**Comment by skreborn** [comment]
@tproenca There have been a [pull request](https://github.com/angular/angular/pull/19733) issued 3 days ago by the author of this issue that aims to fix this. You can find a link to it just above your comment.

**Comment by ehsaanwelcome** [comment]
any update on this?

**Comment by skreborn** [comment]
@ehsaanwelcome The changes in the linked [pull request](https://github.com/angular/angular/pull/19733) were [approved](https://github.com/angular/angular/pull/19733#pullrequestreview-75740481) just 2 hours ago. It should be merged in soon.

**Comment by ehsaanwelcome** [comment]
@skreborn This bug is fixed in 5.0.2 but i still have same issue after upgrading to 5.0.2
first child with query params 'hello=world'
then navigate to another component with queryParams=null and queryParamshandling='merge'

**Comment by mkennedy3000** [comment]
@ehsaanwelcome in order to remove a specific queryParam you need to specify it (`queryParms = { hello: null }` instead of `queryParams = null`).

To remove all queryParams you can set `queryParams=null` or `queryParams={}` without `queryParamsHandling = 'merge'`.

`queryParamsHandling = 'merge'` works the same way as `queryParams = Object.assign(previousQueryParams, newQueryParams)`. So `Object.assign({ hello: 'world' }, null)` is still `{ hello: 'world'}`. I hope that helps you understand how it works.
